File: DynamicProgramming/src/dyn/edit_distance.js
// edit_distance.js
// JavaScript edit distance calculator
// Uses a dictionary for memoization, and Levenshtein's algorithm

// ported from my Python code
// may suffer from Javascript inefficiencies, etc.
max = Math.max, min = Math.min;
ld = function (a, b) {
    var memo = {};
    var step = function (i, j) {
        var k = [i, j];
        if (memo[k] === undefined) {
            memo[k] = min(i, j) === 0 ? max(i, j) : min(
            step(i - 1, j) + 1, // \pm
            step(i, j - 1) + 1, // \mp
            step(i - 1, j - 1) + (a[i - 1] === b[j - 1] ? 0 : 1) // sub
            );
        }
        return memo[k];
    }
    return step(a.length, b.length);
}

File: DynamicProgramming/src/dyn/edit_distance.py
#!/usr/bin/env python

# edit_distance.py
# Python edit distance calculator
# Uses a dictionary for memoization, and Levenshtein's algorithm

def ld(a, b):
  '''
  Accepts two words `a' and `b' and finds the Levenshtein distance
  (edit distance) between them.
  '''
  memo = {}
  def step(i, j):
    if (i, j) not in memo:
      memo[(i, j)] = max(i, j) if min(i, j) is 0 else min(
        step(i-1, j) + 1,                                 # first word:  \pm
        step(i, j-1) + 1,                                 # second word: \mp
        step(i-1, j-1) + (0 if a[i-1] is b[j-1] else 1))  # substitution
    return memo[(i, j)]
  return step(len(a), len(b))
